Refactoring PartialRenderer

tenderlove opened the pull request:
Sorry this diff is so big.

The complexity in `PartialRenderer` has been bothering me for a while, so I decided to do something about it.  When someone calls `render` inside a view, they can pass many different options, and the `PartialRenderer` handled all of them.  I tried to make a table of the different scenarios, and what it would do:

💚 = Provided by user
🤍 = Not Provided by user

Partial Name | Collection | Object | Behavior
-------------|------------|-------|---------
💚                | 🤍               | 🤍     | A regular partial is rendered
💚                | 💚               | 🤍     | Each object in the collection is provided as a local to the specified partial, and the partial is rendered `collection.size` times
💚                | 🤍               | 💚     | The specified partial is rendered with the object as a local
🤍                | 💚               | 🤍     | Each object in the collection is provided as a local, but the partial to be rendered is derived based on the `to_partial_path` of the object
🤍               | 🤍               | 💚      | The specified object is provided as a local, but the partial to be rendered is derived from `to_partial_path`
💚                | 💚               | 💚     | ❌ Not handled, doesn't make sense
🤍                | 💚               | 💚     | ❌ Not handled, doesn't make sense
🤍               | 🤍               | 🤍     | ❌ Not handled, doesn't make sense

Before this PR, `PartialRenderer` would handle all of the above cases (besides the ❌ ones).  This PR extracts two classes from `PartialRenderer`, `ObjectRenderer` and `CollectionRenderer`.

As the name indicates, `CollectionRenderer` is responsible for rendering collections and `ObjectRenderer` responsible for objects.  This lets us simplify the implementation of each class making them easier to understand (and hopefully easier to optimize).  To do this, I had to push some complexity up the stack which you can see [here](https://github.com/rails/rails/blob/8d539bde7b7bd2d2347b804f935f7df7a7ce2e15/actionview/lib/action_view/renderer/renderer.rb#L67).  We have to decompose the `options` hash earlier so we can decide which renderer to use.

I wrote a benchmark so we can try to make sure performance doesn't change (or hopefully gets better!).  It's kind of long, so I put it [here](https://gist.github.com/tenderlove/8be9006e1d410398afdad30f4657e666).

Here are the IPS results on master:

```
Warming up --------------------------------------
      shallow render     4.553k i/100ms
         deep render   172.000  i/100ms
  shallow render obj     3.527k i/100ms
     deep render obj   129.000  i/100ms
   collection render   586.000  i/100ms
Calculating -------------------------------------
      shallow render     47.649k (± 2.5%) i/s -    241.309k in   5.067458s
         deep render      1.730k (± 2.7%) i/s -      8.772k in   5.073434s
  shallow render obj     36.429k (± 2.4%) i/s -    183.404k in   5.037446s
     deep render obj      1.299k (± 2.8%) i/s -      6.579k in   5.069465s
   collection render      5.810k (± 1.4%) i/s -     29.300k in   5.044194s
```

On this branch:

```
Warming up --------------------------------------
      shallow render     4.920k i/100ms
         deep render   184.000  i/100ms
  shallow render obj     3.666k i/100ms
     deep render obj   134.000  i/100ms
   collection render   530.000  i/100ms
Calculating -------------------------------------
      shallow render     50.832k (± 2.9%) i/s -    255.840k in   5.037422s
         deep render      1.849k (± 2.5%) i/s -      9.384k in   5.078979s
  shallow render obj     37.859k (± 2.6%) i/s -    190.632k in   5.038759s
     deep render obj      1.334k (± 2.7%) i/s -      6.700k in   5.026888s
   collection render      5.317k (± 1.5%) i/s -     27.030k in   5.085151s
```

Rendering partials got slightly faster, while rendering collections is slightly slower.  I am confident we can return collection rendering to the same speed or faster with this architecture though. I will make follow up PRs to do so, but I wanted to get this posted sooner rather than later.

tenderlove committed:
pull template keys up

tenderlove committed:
extract template keys to a method

tenderlove committed:
collection with/without template are more similar

tenderlove committed:
collection with/without template are the same

tenderlove committed:
block does not need to be an ivar

tenderlove committed:
Don't rely on the collection ivar

We're going to turn this in to a more complex iterator

tenderlove committed:
variable ivar is not needed anymore

tenderlove committed:
remove useless conditional

tenderlove committed:
Find the layout earlier

Remove duplication in finding a layout.  Remove `@as` ivar.

tenderlove committed:
Don't to_a things when we don't need to.

If the partial renderer is passed a relation or collection proxy, it
will be good for us to resolve that relation or collection proxy as late
as possible.

tenderlove committed:
rely less on the collection ivar

tenderlove committed:
Only iterate over collections once when a path is provided

We'll later reuse this object to speed up the case when a relation is
passed in.

tenderlove committed:
pull iterator allocation in to a factory method

tenderlove committed:
a collection is never rendered with an object

tenderlove committed:
collection rendering has its own class

tenderlove committed:
remove collection references from the superclass

tenderlove committed:
remove more collection references from the superclass

tenderlove committed:
remove at ivar again

tenderlove committed:
introduce an object renderer

tenderlove committed:
remove has_object ivar

tenderlove committed:
partial renderer is specialized for rendering partials now

tenderlove committed:
as: doesn't make sense without an object parameter

The local variable name in `as:` may not be a valid local variable name,
but if there is no object specified to be assigned to the parameter,
then why supply the `as:`?  This commit adds an object for the as param

tenderlove committed:
remove the as_variable method

tenderlove committed:
remove extra ivar set

tenderlove committed:
Move options ivar assignment in to the superclass

tenderlove committed:
move locals and details ivar set in to initialize

tenderlove committed:
Move `partial_path` to ObjectRendering

Partial paths are only required for deriving template info for rendering
objects, which is only the case for the object renderer and the
collection renderer

tenderlove committed:
path ivar is not required

tenderlove committed:
Merging path prefixes is only done in object / collection rendering

tenderlove committed:
Make extracting details cheaper

Details are only used if the user specifies a format or local, etc when
rendering a template.  This is a fairly uncommon thing to do, so lets
make it cheaper when there are no details specified

tenderlove committed:
move object specific stuff in to the object processing code

tenderlove committed:
We don't need this private method

tenderlove committed:
Move renderers in to their own files

We don't need to keep all of these in the partial renderer file

tenderlove committed:
remove collection ivar

tenderlove committed:
Remove conditional

Only one case can possibly be missing a partial.  We can move the cache
check to that branch and eliminate the conditional

tenderlove committed:
Reduce iterations for non-homogeneous collections

We only need to iterate over the paths twice (this was doing it 3 times)

tenderlove committed:
always lookup template from path cache

tenderlove committed:
fix require

tenderlove committed:
Move around local variable derivation

This reduces object allocations in collection rendering

tenderlove committed:
make the cops happy

tenderlove commented:
Btw, I found an interesting thing.  It's OK for you to have a partial that yields, for example:

```erb
Hello <%= yield %>
```

But it doesn't work in all cases.  For example:

```ruby
# frozen_string_literal: true

require "active_record"
require "active_record/railties/collection_cache_association_loading"
require "action_controller"
require "action_view"
require "tmpdir"
require "benchmark/ips"

ActionView::PartialRenderer.prepend(ActiveRecord::Railties::CollectionCacheAssociationLoading)

class TestController < ActionController::Base
  def _prefixes; ["."]; end # Search the current directory rather than "test/"
end

class Neat
  def to_partial_path; "thing"; end
end

Dir.mktmpdir do |dir|
  TestController.view_paths = dir

  File.write File.join(dir, "_thing.html.erb"), "Hello <%= yield %>"

  cv = TestController.new.view_context

  obj        = Neat.new
  collection = [Neat.new] * 50

  cv.render("thing")    { "World" }
  cv.render(obj)        { "World" }

  begin
    cv.render(collection) { "World" } # Exception
  rescue ActionView::Template::Error
    puts "Oh no!"
  end
end
```

You can render the partial via a string, or via an object that responds to `to_partial_path`, but if you have a collection of objects that return that partial, you get an exception.  The fix is pretty trivial, and I have a patch for it locally (against this branch).  I just thought it was an interesting edge case that apparently nobody has seen?

kaspth commented:
Love all this ❤️ I've only used `yield` in templates a few times, and I've been slightly surprised they worked, but never with collections. Nice find!

tenderlove committed:
Merge branch 'master' into collection-refactor

* master:
  Add a regression test that ActionText caught
  [ci skip] Use yml extension for locale files
  Fix `helper_method` in `ActionView::TestCase` to allow keyword arguments
  Fix `delegate_missing_to` to allow keyword arguments
  Dump the schema or structure of a database when calling db:migrate:name
  Reset the `ActiveRecord::Base` connection after `rails db:migrate:name`
  Fix `unscope` when an `eq` node which has no arel attribute
  Remove unused argument
  Disallow calling `connected_to` on subclasses of `ActiveRecord::Base`
  More less and lazy allocation for `assign_attributes` and `_assign_attributes`
  Tweak contributing_to_ruby_on_rails.md [ci skip]
  Clarify the difference between (old) `spec_name` and `connection_specification_name`
  Remove duplicate part from deprecation warning
  Fix deprecation warnings in connection_handlers_sharding_db_test.rb
  Fixup CHANGELOGs [ci skip]
  `reset_column_information` does not reset @predicate_builder
  Simplify FixtureResolver to reuse filtering logic
  Mostly remove bad test
  Use type attribute in ActionView::Helpers::JavaScriptHelper#javascript_tag example
  Update some references to finder options [ci skip]

tenderlove committed:
Fix issue in ActionText

ActionText was using `render` in a way that ActionView didn't have tests
for.  This is a fix for it.

tenderlove committed:
pull preloading behavior in to the collection renderer

tenderlove committed:
remove usless attr_reader

rafaelfranca commented on actionview/lib/action_view/renderer/collection_renderer.rb:
```diff
@@ -0,0 +1,188 @@
+      end
+    end
+
+    class PreloadCollectionIterator < SameCollectionIterator # :nodoc:
```
Do you think it is better to have Active Record specific behavior in Action View? I can clearly see how this improved the code but at some time this is the first time we have Active Record specific behavior inside Action View. It will also call the conditional every time a collection is passed even if the collection is the an Active Record collection.

One alternative is extract the conditional:

```
      collection = if collection.respond_to?(:preload_associations)
        PreloadCollectionIterator.new(collection, partial, iter_vars, collection)
      else
        SameCollectionIterator.new(collection, partial, iter_vars)
      end
```

to a method and only override that method if Active Record is present.

I'm not strong about this either so I'll go with what you decided. I like the way you implemented this as well.

tenderlove replied:
> Do you think it is better to have Active Record specific behavior in Action View? I can clearly see how this improved the code but at some time this is the first time we have Active Record specific behavior inside Action View.

I thought about that, but we're not referencing any constants from Active Record, only checking to see if the collection responds to `preload_associations`. Any collection could technically support preloading (though I guess  only Active Record does 😬).

My thought process was basically:

* ActionView is usually used with ActiveRecord, so most people won't pay a penalty (this is also why I included the module [in the benchmark](https://gist.github.com/tenderlove/8be9006e1d410398afdad30f4657e666#file-render_benchmark-rb-L10)
* Collection rendering is probably not a bottleneck in apps that use AV but don't use AR, so the extra conditional probably doesn't matter
* Maybe other collection types would like the opportunity to implement preloading, and this decouples us from ActiveRecord specifically

Anyway, I don't have a strong preference, but I do like that we could remove that module, the initializer, and decouple from AR (kind of).

rafaelfranca replied:
👍

yashLadha commented on actionview/lib/action_view/renderer/abstract_renderer.rb:
```diff
@@ -27,6 +29,86 @@ def render
+          path = if object.respond_to?(:to_partial_path)
+            object.to_partial_path
+          else
+            raise ArgumentError.new("'#{object.inspect}' is not an ActiveModel-compatible object. It must implement :to_partial_path.")
```
Line length is long, can be reduced to a function call to render the string.

kaspth commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -29,15 +35,15 @@ def cache_collection_render(instrumentation_payload, view, template)
         #
         # If the cache is missing elements then
         # the block will be called against the remaining items
         # in the @collection.
```
Looks like the documentation is now slightly off here and elsewhere in the file. Could update or maybe scrap? Think it blurs more than helps?

tenderlove replied:
I'll remove it.  Now that the caller is required to use the collection that's yielded to the block, I don't think we need this documentation (the ivar isn't mutated anymore).

kaspth reviewed (approved):
🙌

kaspth commented:
@tenderlove looks like there's the docs removal pending, but other than that should be good to get this first version in?

tenderlove commented:
@kaspth yep, thanks for pinging me.  I'll clean up the docs and merge this today

tenderlove committed:
Remove out-of-date comments

kaspth commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -29,15 +35,9 @@ def cache_collection_render(instrumentation_payload, view, template)
         # Extract the items for the keys that are not found
         # Set the uncached values to instance variable @collection
```
@tenderlove pst, this doc too 😄

tenderlove replied:
oops, missed one! 😬

tenderlove commented:
hrm.  I don't think these failures are my fault.  I'm going to merge anyway
